rpki-validator/src/main/java/net/ripe/rpki/validator3/background/Throttled.java

```diff
@@ -44,85 +43,104 @@
         Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors());
 
-    public Throttled(long minIntervalInSeconds) {
-        this.minIntervalInSeconds = minIntervalInSeconds;
+    public Throttled(long minIntervalMs) {
+        this.minIntervalMs = minIntervalMs;
     }
 
     public void trigger(Key key, Runnable r) {
-        final Runnable wrappedRunnable = () -> {
+        synchronized (actionMap) {
+            final Instant now = Instant.now();
+            final Action action = actionMap.get(key);
+
+            if (action == null) {
+                actionMap.put(key, Action.toBeExecutedASAP(now, r));
+                scheduledExecutor.execute(actionRunner(key));
+            } else if (action.running) {
+                actionMap.put(key, action.scheduled(r));
+            } else if (action.alreadyScheduled) {
+                actionMap.put(key, action.replaced(r));
+            } else {
+                final Instant lastTime = action.getExecutionTime();
+                final long delay = minIntervalMs - (now.toEpochMilli() - lastTime.toEpochMilli());
+                actionMap.put(key, action.scheduled(r));
+                scheduledExecutor.schedule(actionRunner(key), delay, TimeUnit.MILLISECONDS);
+            }
+        }
+    }
+
+    private Runnable actionRunner(Key key) {
+        return () -> {
             try {
-                r.run();
-            } finally {
-                final Instant after = Instant.now();
+                Action action;
                 synchronized (actionMap) {
-                    actionMap.put(key, Action.toBeExecutedSomeTimeLater(after));
+                    action = actionMap.get(key);
+                    actionMap.put(key, action.started(Instant.now()));
                 }
-            }
-        };
-
-        final Instant now = Instant.now();
-        synchronized (actionMap) {
-            final Action action = actionMap.get(key);
-            if (action != null) {
-                if (!action.alreadyScheduled) {
-                    final Instant lastTime = action.getExecutionTime();
-                    final Duration between = Duration.between(lastTime, now);
-                    if (between.getSeconds() < minIntervalInSeconds) {
-                        final long delay = minIntervalInSeconds - between.getSeconds();
-                        actionMap.put(key, action.scheduled());
-                        scheduledExecutor.schedule(wrappedRunnable, delay, TimeUnit.SECONDS);
-                    } else {
-                        actionMap.put(key, action.scheduled());
-                        scheduledExecutor.execute(wrappedRunnable);
+                action.runnable.run();
+            } finally {
+                synchronized (actionMap) {
+                    Action action = actionMap.get(key);
+                    actionMap.put(key, action.stopped());
+                    if (action.alreadyScheduled) {
+                        trigger(key, action.runnable);
```

Comment:
And why do we trigger action for the same key again? To process one if it was waiting on the `synchronized`?

Reply:
To process actions that were scheduled while the action was running. So we wait until the running action finishes to re-trigger the action, taking into account the `minIntervalMs`.